src/ess/reduce/nexus.py

```diff
@@ -0,0 +1,329 @@
+        on its NeXus class.
+
+    Returns
+    -------
+    :
+        A data group containing all data stored in
+         the source NeXus group.
+    """
+    return RawSource(
+        _load_group_with_positions(
+            file_path,
+            group_name=source_name,
+            nx_class=snx.NXsource,
+            instrument_name=instrument_name,
+        )
+    )
+
+
+def load_sample(
+    file_path: Union[FilePath, NeXusFile, NeXusGroup],
+) -> RawSample:
+    """Load a sample from a NeXus file.
+
+    The sample is located based on its NeXus class.
+    There can be only one sample in a NeXus file or
+    in the group given as ``file_path``.
+
+    Parameters
+    ----------
+    file_path:
+        Indicates where to load data from.
+        One of:
+
+        - Path to a NeXus file on disk.
+        - File handle or buffer for reading binary data.
+        - A ScippNexus group of the root of a NeXus file.
+
+    Returns
+    -------
+    :
+        A data group containing all data stored in
+         the sample NeXus group.
+    """
+    with _open_nexus_file(file_path) as f:
+        entry = f['entry']
+        loaded = cast(sc.DataGroup, _unique_child_group(entry, snx.NXsample, None)[()])
+    return RawSample(loaded)
+
+
+def _load_group_with_positions(
+    file_path: Union[FilePath, NeXusFile, NeXusGroup],
+    *,
+    group_name: Optional[str],
+    nx_class: Type[snx.NXobject],
+    instrument_name: Optional[InstrumentName] = None,
+) -> sc.DataGroup:
+    with _open_nexus_file(file_path) as f:
+        entry = f['entry']
+        instrument = _unique_child_group(entry, snx.NXinstrument, instrument_name)
+        loaded = cast(
+            sc.DataGroup, _unique_child_group(instrument, nx_class, group_name)[()]
+        )
+        loaded = snx.compute_positions(loaded)
+        return loaded
+
+
+def _open_nexus_file(
+    file_path: Union[FilePath, NeXusFile, NeXusGroup]
+) -> ContextManager:
+    if isinstance(file_path, getattr(NeXusGroup, '__supertype__', type(None))):
+        return nullcontext(file_path)
+    return snx.File(file_path)
+
+
+def _unique_child_group(
+    group: snx.Group, nx_class: Type[snx.NXobject], name: Optional[str]
+) -> snx.Group:
+    if name is not None:
+        child = group[name]
+        if isinstance(child, snx.Field):
+            raise ValueError(
+                f"Expected a NeXus group as item '{name}' but got a field."
+            )
+        if child.nx_class != nx_class:
+            raise ValueError(
+                f"The NeXus group '{name}' was expected to be a "
+                f'{nx_class} but is a {child.nx_class}.'
+            )
+        return child
+
+    children = group[nx_class]
+    if len(children) != 1:
+        raise ValueError(f'Expected exactly one {nx_class} group, got {len(children)}')
+    return next(iter(children.values()))  # type: ignore[return-value]
+
+
+def extract_detector_data(
+    detector: RawDetector, detector_name: DetectorName
```

Comment:
```suggestion
    detector: RawDetector
```
I don't think we should have `DetectorName` here. Instead return the event data, if found, otherwise the child data array (dense data). Reason: I don't think we should rely on the detector name being related to the name of the event-data subgroup. In the case of monitors there may be no subgroup at all. Also it is simpler to just look for the child data array(s).

Reply:
How can I identify the data? It's just one of many data arrays in the group.

Reply:
Or can there only be one data array per detector / monitor?